### PR TITLE
Compare user based on uid

### LINKED ISSUE #3186: user comparison based on uid
We are currently comparing user objects based on the loginname:

https://github.com/privacyidea/privacyidea/blob/0f085fd2ed59cd4567989c9bbe55ea01c29ced3a/privacyidea/lib/user.py#L150

I think this is due to the fact, that in some cases a user object could not be readily initialized and actually have no uid - or it was this way.

We are observing that we could run into case sensitivity problems, depending on the scenario, what a user enters as his username and what the resolver returns as a username. So I think we should compare user objects based on the uid, if it is available.

The following code should provide us backward compatibility (if there is a userobject without a uid) and also compare via uid, if it exists.

~~~~python
    def __eq__(self, other):
        """
        Compare two User Objects.

        :param other: The other User object, to which this very object is
        compared.
        :type other: User object
        :return: True or False
        :rtype: bool
        """
        if not isinstance(other, type(self)):
            log.info("comparing users of different classes.")
            return False
        if (self.resolver != other.resolver) or (self.realm != other.realm):
            log.info("Users are not in the same resolver and realm")
            return False
        if self.uid and other.uid:
            log.info("comparing based on uid")
            return self.uid == other.uid
        log.info("comparing based on login")
        return self.login == other.login
~~~~

We might add this to a 3.7.2.

### PR DESCRIPTION
**Pull request by cornelinux**
Fixes #3186

### CODE REVIEW

**Comment by cornelinux** [comment]
> Hello @cornelinux! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:
> 
> * In the file [`tests/test_lib_user.py`](https://github.com/privacyidea/privacyidea/blob/ae2e52457a136d5ece69cddbaceb61249ffad6f0/tests/test_lib_user.py):
> 
> > [Line 507:51](https://github.com/privacyidea/privacyidea/blob/ae2e52457a136d5ece69cddbaceb61249ffad6f0/tests/test_lib_user.py#L507): [E711](https://duckduckgo.com/?q=pep8%20E711) comparison to None should be 'if cond is None:'

I am doing `== None`  on purpose, to actually test the `__eq__` method.

**Review by plettich — CHANGES_REQUESTED** [review]

File: tests/test_lib_user.py [hunk 1]
```diff
@@ -476,6 +476,45 @@ def test_18_user_with_several_phones(self):
...
+        rid = save_resolver(params)
+        self.assertTrue(rid > 0)
+        (added, failed) = set_realm("ldap", ["ldapresolver"])
+        self.assertEqual(len(added), 1)
+        self.assertEqual(len(failed), 0)
+
+        # Comparing different objects fails
+        self.assertFalse(User("salesman", "ldap") == None)
+        # corparing different real or resolver fails
+        self.assertFalse(User("root", self.realm1) == User("salesmann", "ldap"))
```
[the remark this answers is not in the document]
Comment (cornelinux) [reply tests/test_lib_user.py]: thx. done in ceca6a97

File: privacyidea/lib/user.py [hunk 2]
```diff
@@ -147,8 +147,17 @@ def __eq__(self, other):
...
+        if not isinstance(other, type(self)):
+            log.info("Comparing a non-user object.")
+            return False
+        if (self.resolver != other.resolver) or (self.realm != other.realm):
+            log.info("Users are not in the same resolver and realm")
+            return False
+        if self.uid and other.uid:
+            log.info(u"Comparing based on uid: {0!s} vs {1!s}".format(self.uid, other.uid))
+            return self.uid == other.uid
+        log.info(u"Comparing based on login: {0!s} vs {1!s}".format(self.login, other.login))
```
[the remark this answers is not in the document]
Comment (cornelinux) [reply privacyidea/lib/user.py:159]: OK, will downgrade this to debug.

[the remark this answers is not in the document]
Comment (cornelinux) [reply privacyidea/lib/user.py]: Done in eb225a8f

**Review by plettich — APPROVED** [review]

**Pull request merged**